content/id-datatype/index.md

```diff
@@ -0,0 +1,103 @@
+---
+rfc:
+start_date: 2018-08-02
+pr: openregister/registers-rfcs#22
+status: draft
+---
+
+# ID datatype
+
+## Summary
+
+This RFC proposes a new datatype to constrain and govern identifiers for
+records.
+
+
+## Motivation
+
+Currently, identifiers (i.e. keys) are defined as of any datatype available.
+Usage shows only strings have been used thus far.
+
+Allowing any datatype imposes a hard task on the user trying to validate
+what to expect in there. Other questions arise when you analyise how each
+datatype is expected to behave in all contexts identifiers are used.
+
+Questions like “What are the rules for an identifier when used as part of a
+CURIE?”, “What does it mean to have a String (full UTF-8 range)?”.
+
+Notice that [issue 18](https://github.com/openregister/registers-rfcs/issues/18)
+has a discussion related to this RFC.
+
+This RFC aims to improve the situation by proposing a new ID datatype
+constrained enough to make validation viable, that works well with CURIEs and
+overall that guides users without preventing valid use cases.
+
+
+## Explanation
+
+The new ID datatype is a subset of the String datatype. It helps keeping
+backwards compatibility and builds on top of how Registers are used in
+reality.
+
+Evidence says that identifiers are composed primarily by alphanumeric ASCII
+characters with a few other printable characters:
+
+* `-` (U+002D)
+* `_` (U+005F)
+* `.` (U+002E)
+* `/` (U+002F)
+
+Note that `/` has meaning in URLs and hence it has meaning in CURIEs. This
+means that a CURIE must always percent-encode this character as `%2F`.
```

Comment:
Does this mean that a valid curie would be "country:a%2Fb" and it would resolve to https://country.register.gov.uk/records/a%2Fb? Whereas an invalid curie would be "country:a/b" as this would resolve to https://country.register.gov.uk/records/a/b? 

My question is whether we could allow "/" in curies (such that "country:a/b" is valid) but dictate that it is the responsibility of those interpreting a curie to resolve it to https://country.register.gov.uk/records/a%2Fb? However, I guess this would restrict us from being able to interpret "country:a/b" as https://country.register.gov.uk/records/a/b in the future if we wanted to. So I think it makes sense to either encode it in the curie or drop them altogether for now.

Given that the industrial-classification registers are still in alpha, we will need to either re-load them to encode the `/` characters in their IDs or we could use new keys. I'm not sure what the status of these registers are.

Reply:
> Does this mean that a valid curie would be "country:a%2Fb" and it would resolve to https://country.register.gov.uk/records/a%2Fb? Whereas an invalid curie would be "country:a/b" as this would resolve to https://country.register.gov.uk/records/a/b?

Correct.

> My question is whether we could allow "/" in curies (such that "country:a/b" is valid) but dictate that it is the responsibility of those interpreting a curie to resolve it to https://country.register.gov.uk/records/a%2Fb? However, I guess this would restrict us from being able to interpret "country:a/b" as https://country.register.gov.uk/records/a/b in the future if we wanted to. So I think it makes sense to either encode it in the curie or drop them altogether for now.

Yes, if we diverge from the standard definition of how a CURIE works, it means we will have to define the rules to expand to something like https://country.register.gov.uk/records/a/b

I rather stick to CURIEs and try to minimise the use of `/` in identifiers.

> Given that the industrial-classification registers are still in alpha, we will need to either re-load them to encode the / characters in their IDs or we could use new keys. I'm not sure what the status of these registers are.

Umm, I need to address the ID definition because as IDs it is valid to use "a/b" but when they are part of a CURIE or a URL, they need to be encoded.